---
title: "pds08_3"
author: "Braden T Tierney"
date: "10/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(ggplot2)
library(umap)
library(cowplot)
library(vegan)
library(reshape2)
library(ggrepel)
library(broom)
library(ggpubr)

theme_set(theme_cowplot())
```


```{r}
setwd('~/Desktop/google_drive/My Drive/pds08/')

bracken = read.csv('bracken_output/merged_bracken_output.tsv',sep='\t',header=T)

#filter out rare bugs
bracken_num = bracken %>% select(name,all_of(colnames(bracken)[!grepl('frac',colnames(bracken))])) %>% column_to_rownames('name') %>% select(-taxonomy_id,-taxonomy_lvl)
orgs_to_keep = rowMeans(bracken_num) %>% data.frame %>% rownames_to_column('organism') %>% filter(.>100) %>% select(organism) %>% unlist %>% unname

bracken = bracken %>% filter(name %in% orgs_to_keep) %>% select(name,taxonomy_lvl,all_of(colnames(bracken)[grepl('frac',colnames(bracken))]))
colnames(bracken) = gsub('\\.bracken_frac','',colnames(bracken))

# get samples to keep 
basekeep = colnames(bracken)[grepl('_01',colnames(bracken))] 
basekeep = gsub('_01','',basekeep)
endkeep = colnames(bracken)[grepl('_02',colnames(bracken))] 
endkeep = gsub('_02','',endkeep)

samples_to_keep = intersect(basekeep,endkeep)

colnames(bracken) = gsub('_01','_baseline',colnames(bracken))
colnames(bracken) = gsub('_02','_endpoint',colnames(bracken))

bracken_s = bracken %>% filter(taxonomy_lvl=='S') %>% select(-taxonomy_lvl)

metadata = readRDS('pds08_metadata.rds') # %>% filter(b_bm_weekly<=4.2)
#demarcate responder vs nonresponder
metadata = metadata %>% mutate(RESP_STATUS = as.factor(if_else(rx != 'Treatment',"PLACEBO",if_else(d_bm_weekly>=2,'RESPONDER',if_else(abs(d_bm_weekly)<2,'NOCHANGE-TREATMENT','WORSE-TREATMENT')))))
metadata = metadata %>% mutate(RESP_STATUS = as.factor(if_else(RESP_STATUS != 'PLACEBO',as.character(RESP_STATUS),if_else(d_bm_weekly>=2,'IMPROVED-PLACEBO',if_else(abs(d_bm_weekly)<2,'NOCHANGE-PLACEBO','WORSE-PLACEBO')))))
metadata = metadata %>% mutate(IMPROVED = if_else(RESP_STATUS == 'IMPROVED-PLACEBO' | RESP_STATUS == 'RESPONDER','IMPROVED',if_else(RESP_STATUS == 'NOCHANGE-PLACEBO' | RESP_STATUS == 'NOCHANGE-TREATMENT','NOCHANGE','WORSE')))

#compute diversity and richness
data = bracken_s %>% column_to_rownames('name')

toremove = colnames(data) %>% strsplit('_') %>% map_chr(1) %>% table %>% data.frame %>% filter(Freq!=2) %>% select(colnames(.)[1]) %>% unlist %>% unname
data = data %>% data.frame%>% select(!(grep(paste(toremove, collapse="|"), colnames(data))))

#data = data[rowMeans(data)>.00001,] 
shannon = map(colnames(data), function(x) diversity(data[,x],index='shannon')) %>% unlist %>% unname
simpson = map(colnames(data), function(x) diversity(data[,x],index='simpson')) %>% unlist %>% unname
richness = map(colnames(data), function(x) length(which(data[,x]!=0))) %>% unlist %>% unname

divs = data.frame(shannon,simpson,richness)
rownames(divs) = colnames(data)
divs = divs %>% rownames_to_column('Sample_ID')
divs = divs %>% mutate(timepoint = strsplit(Sample_ID,'_')%>% map_chr(2),Sample_ID = strsplit(Sample_ID,'_')%>% map_chr(1)) 

# merge all data into one frame
bracken_s = bracken_s %>% column_to_rownames('name') %>% t %>% data.frame %>% rownames_to_column('Sample_ID') 
bracken_s$timepoint = strsplit(bracken_s$Sample_ID,'_')%>% map_chr(2)
bracken_s$Sample_ID = strsplit(bracken_s$Sample_ID,'_')%>% map_chr(1)

bracken_merged = left_join(bracken_s,divs,by=c('Sample_ID','timepoint'))
abdata = left_join(bracken_merged,metadata,by='Sample_ID') %>% filter(!is.na(RESP_STATUS)) %>% filter(Sample_ID %in% samples_to_keep)

```

```{r}
# metaphlan instead of bracken
setwd('~/Desktop/google_drive/My Drive/pds08/')

baseline_metaphlan = readRDS('metaphlan_baseline.rds') %>% mutate(Sample_ID = paste(Sample_ID,'_baseline',sep=''))
endpoint_metaphlan = readRDS('metaphlan_endpoint.rds') %>% mutate(Sample_ID = paste(Sample_ID,'_endpoint',sep=''))
delta_metaphlan = readRDS('metaphlan_delta.rds') %>% mutate(Sample_ID = paste(Sample_ID,'_delta',sep=''))

data = bind_rows(baseline_metaphlan,endpoint_metaphlan) %>% column_to_rownames('Sample_ID') 
data = data %>% t
#data = data %>% select(!all_of(grep('s__',colnames(data)))) #%>% t
#data = data %>% select(all_of(grep('s__',colnames(data)))) %>% t
#data = data[rowMeans(data)>.0001,] 

toremove = colnames(data) %>% strsplit('_') %>% map_chr(1) %>% table %>% data.frame %>% filter(Freq!=2) %>% select(colnames(.)[1]) %>% unlist %>% unname
data = data %>% data.frame%>% select(!(grep(paste(toremove, collapse="|"), colnames(data))))

data_sub = data[!grepl('o__',rownames(data)),]
data_sub = data_sub[grepl('c__',rownames(data_sub)),]
shannon = map(colnames(data_sub), function(x) diversity(data_sub[,x],index='shannon')) %>% unlist %>% unname
simpson = map(colnames(data_sub), function(x) diversity(data_sub[,x],index='simpson')) %>% unlist %>% unname
richness = map(colnames(data_sub), function(x) length(which(data_sub[,x]!=0))) %>% unlist %>% unname

divs = data.frame(shannon,simpson,richness)
rownames(divs) = colnames(data)
divs = divs %>% rownames_to_column('Sample_ID')
divs = divs %>% mutate(timepoint = strsplit(Sample_ID,'_')%>% map_chr(2),Sample_ID = strsplit(Sample_ID,'_')%>% map_chr(1)) 

# merge all data into one frame
metaphlan_s = data %>% t %>% data.frame %>% select(all_of(grep('s__',colnames(.))))%>% rownames_to_column('Sample_ID') 
metaphlan_s$timepoint = strsplit(metaphlan_s$Sample_ID,'_')%>% map_chr(2)
metaphlan_s$Sample_ID = strsplit(metaphlan_s$Sample_ID,'_')%>% map_chr(1)
metaphlan_merged = left_join(metaphlan_s,divs,by=c('Sample_ID','timepoint'))
abdata = left_join(metaphlan_merged,metadata,by='Sample_ID') %>% filter(!is.na(RESP_STATUS))
```

```{r}

metaphlan_b_e = bind_rows(baseline_metaphlan,endpoint_metaphlan) %>% column_to_rownames('Sample_ID')

metaphlan_b_e_umap = umap(metaphlan_b_e)
umap_output = metaphlan_b_e_umap$layout %>% data.frame %>% rownames_to_column('key') %>% as.tibble %>% mutate(Sample_ID = strsplit(key,'_') %>% map_chr(1),timepoint = strsplit(key,'_') %>% map_chr(2))
umap_output = left_join(umap_output,metadata,by='Sample_ID')  %>% filter(!is.na(RESP_STATUS))

ggplot(data = umap_output ,aes(x=X1,y=X2,group=Sample_ID,color=as.factor(RESP_STATUS))) + geom_point(size=3,aes(shape=timepoint)) + geom_line()

ggplot(data = umap_output %>% filter(timepoint == 'baseline'),aes(x=X1,y=X2,group=Sample_ID,color=as.factor(RESP_STATUS))) + geom_point(size=3,aes(shape=timepoint))

ggplot(data = umap_output %>% filter(timepoint == 'endpoint'),aes(x=X1,y=X2,group=Sample_ID,color=as.factor(RESP_STATUS))) + geom_point(size=3,aes(shape=timepoint))

```


```{r}
# BETA COMP

setwd('~/Desktop/google_drive/My Drive/pds08/')

# baseline
abdata_b = abdata %>% select(Sample_ID,timepoint,RESP_STATUS,b_bm_weekly,b_bloat,b_pain,shannon,simpson,richness,rx,IMPROVED) %>% filter(timepoint=='baseline') %>% mutate(Sample_ID = paste(Sample_ID,timepoint,sep='_'))
abdata_e = abdata %>% select(Sample_ID,timepoint,RESP_STATUS,b_bm_weekly,b_bloat,b_pain,shannon,simpson,richness,rx,IMPROVED) %>% filter(timepoint=='endpoint') %>% mutate(Sample_ID = paste(Sample_ID,timepoint,sep='_'))
abdata_long = bind_rows(abdata_b,abdata_e) %>% column_to_rownames("Sample_ID")

beta_dist = as.matrix(vegdist(metaphlan_b_e,index = "bray"))
beta_dist_b = as.matrix(vegdist(baseline_metaphlan %>% column_to_rownames('Sample_ID'),index = "bray"))
beta_dist_e = as.matrix(vegdist(endpoint_metaphlan %>% column_to_rownames('Sample_ID'),index = "bray"))


pheatmap::pheatmap(mat = beta_dist, show_rownames = F,show_colnames = T,cluster_rows=T,cluster_cols =T,annotation_row = abdata_long %>% select(RESP_STATUS,IMPROVED))

pdf('microbiome_baseline_beta.pdf',width=10,height=10)
pheatmap::pheatmap(mat = beta_dist_b, show_rownames = F,show_colnames = T,cluster_rows=T,cluster_cols =T,annotation_row = abdata_long %>% select(RESP_STATUS,IMPROVED))
dev.off()

pheatmap::pheatmap(mat = beta_dist_e, show_rownames = F,show_colnames = T,cluster_rows=T,cluster_cols =T,annotation_row = abdata_long %>% select(RESP_STATUS,IMPROVED))

melted = melt(beta_dist) %>% mutate(id=strsplit(as.character(Var1),'_') %>% map_chr(1),id2=strsplit(as.character(Var2),'_') %>% map_chr(1)) %>% filter(id == id2,Var1!=Var2) %>% mutate(tp = strsplit(as.character(Var1),'_') %>% map_chr(2)) %>% filter(tp =='baseline')%>% select(id,value) %>% rename(Sample_ID = id)
melted_meta = left_join(melted, metadata %>% select(Sample_ID,RESP_STATUS,IMPROVED))


ggplot(melted_meta %>% arrange(value) %>% filter(!is.na(RESP_STATUS)),aes(x=RESP_STATUS,y=value)) + geom_point(aes(color=RESP_STATUS)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
foo = compare_means(data = melted_meta,value ~ RESP_STATUS,method="t.test")
write.csv(foo,'~/temp.csv')
ggsave('beta_self_dis.pdf',width=6,height=6)
anova(lm(value ~ RESP_STATUS,data=melted_meta))

```

```{r}

#look at diversity between responders and nonresponders

#GENERATE A NEW FIGURE THAT SHOWS THIS FOR ALL PHYLOGENETIC LEVELS
#AND SUPP FIGS THAT SHOW IT WITH BRACKEN FOR ALL LEVELS WITH DIFFERENT READ CUTOFFS

setwd('~/Desktop/google_drive/My Drive/pds08/')

abdata_div = abdata %>%  select(Sample_ID,shannon,simpson,richness,RESP_STATUS,rx,timepoint,IMPROVED)
abdata_div_melted = melt(abdata_div)

ggplot(data = abdata_div_melted, aes(x = as.factor(timepoint), y = value)) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +stat_compare_means(paired = T,method='t.test')+ geom_point(aes(fill=as.factor(timepoint),group=Sample_ID), position = position_dodge(0.2)) + geom_line(aes(group=Sample_ID),color='grey',position = position_dodge(0.2)) + facet_grid(cols=vars(RESP_STATUS),rows=vars(variable),scales='free')
ggsave('~/rnr_div.pdf',width=15,height=15)

ggplot(data = abdata_div_melted %>% filter(timepoint == 'baseline'), aes(x = as.factor(RESP_STATUS), y = value)) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ geom_point(aes(fill=as.factor(timepoint),group=Sample_ID), position = position_dodge(0.2)) + facet_grid(rows=vars(variable),scales='free')
ggsave('~/rnr_div_base.pdf',width=15,height=15)

compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='richness') ,value ~ RESP_STATUS,method='t.test')
compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='shannon') ,value ~ RESP_STATUS,method='t.test')
compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='simpson') ,value ~ RESP_STATUS,method='t.test')

ggplot(data = abdata_div_melted, aes(x = as.factor(timepoint), y = value)) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +stat_compare_means(paired = T,method='t.test')+ geom_point(aes(fill=as.factor(timepoint),group=Sample_ID), position = position_dodge(0.2)) + geom_line(aes(group=Sample_ID),color='grey',position = position_dodge(0.2)) + facet_grid(cols=vars(IMPROVED),rows=vars(variable),scales='free')

ggplot(data = abdata_div_melted %>% filter(timepoint == 'baseline'), aes(x = as.factor(IMPROVED), y = value)) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ geom_point(aes(fill=as.factor(timepoint),group=Sample_ID), position = position_dodge(0.2)) + facet_grid(rows=vars(variable),scales='free')

compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='richness') ,value ~ IMPROVED,method='t.test')
compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='shannon') ,value ~ IMPROVED,method='t.test')
compare_means(data = abdata_div_melted %>% filter(timepoint == 'baseline') %>% filter(variable=='simpson') ,value ~ IMPROVED,method='t.test')

```

```{r}
# look for bifidogenic effect
setwd('~/Desktop/google_drive/My Drive/pds08/')
bifidos = colnames(abdata)[grepl('Bifido',colnames(abdata))]

abdata_bif = abdata %>%  select(Sample_ID,all_of(bifidos),RESP_STATUS,rx,timepoint)
abdata_bif_melted = melt(abdata_bif) 
abdata_bif_melted$variable = strsplit(as.character(abdata_bif_melted$variable),"\\.s__") %>% map_chr(2)

ggplot(data = abdata_bif_melted, aes(x = as.factor(timepoint), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + geom_line(aes(group=Sample_ID),color='grey') + facet_grid(cols=vars(RESP_STATUS),rows = vars(variable),scales='free')+theme(strip.text.y = element_text(angle = 0))+ stat_compare_means(paired=T,method='t.test')
ggsave('bifido_change_across_tp.pdf',width=30,height=40)

ggplot(data = abdata_bif_melted %>% filter(timepoint == 'baseline'), aes(x = as.factor(RESP_STATUS), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + facet_grid(rows = vars(variable),scales='free')+theme(strip.text.y = element_text(angle = 0))
ggsave('bifido_baseline.pdf',width=20,height=15)
compare_means(data = abdata_bif_melted %>% filter(timepoint == 'baseline') ,value ~ RESP_STATUS)
ggplot(data = abdata_bif_melted %>% filter(timepoint == 'endpoint'), aes(x = as.factor(RESP_STATUS), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + facet_grid(rows = vars(variable),scales='free')+theme(strip.text.y = element_text(angle = 0))
compare_means(data = abdata_bif_melted %>% filter(timepoint == 'endpoint') ,value ~ RESP_STATUS)
 ggsave('bifido_endpoint.pdf',width=30,height=15)

###OVERALL
abdata_bif_melted_all_bif =abdata_bif_melted %>% select(Sample_ID,RESP_STATUS,timepoint,value,rx) %>% group_by(Sample_ID,timepoint,RESP_STATUS) %>% mutate(value=sum(value))
ggplot(data = abdata_bif_melted_all_bif, aes(x = as.factor(timepoint), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + geom_line(aes(group=Sample_ID),color='grey') + facet_grid(cols=vars(RESP_STATUS)) + stat_compare_means(paired=T,method='t.test')
ggsave('bifido_change_overall.pdf',width=20,height=15)
ggplot(data = abdata_bif_melted_all_bif %>% filter(timepoint == 'baseline'), aes(x = as.factor(RESP_STATUS), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point()
ggsave('bifido_base_overall.pdf',width=20,height=15)
compare_means(data = abdata_bif_melted_all_bif %>% filter(timepoint == 'baseline') ,value ~ RESP_STATUS)
ggplot(data = abdata_bif_melted_all_bif %>% filter(timepoint == 'endpoint'), aes(x = as.factor(RESP_STATUS), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point()
ggsave('bifido_end_overall.pdf',width=30,height=15)
compare_means(data = abdata_bif_melted_all_bif %>% filter(timepoint == 'endpoint') ,value ~ RESP_STATUS)

### REPEAT BY RX

ggplot(data = abdata_bif_melted, aes(x = as.factor(timepoint), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + geom_line(aes(group=Sample_ID),color='grey') + facet_grid(cols=vars(rx),rows = vars(variable),scales='free')+theme(strip.text.x = element_text(angle = 0)) +theme(axis.text.y = element_text(angle = 90, vjust = 0.5, hjust=1)) + stat_compare_means(paired=T,method='t.test')
ggsave('bifido_rx_timepoint.pdf',width=10,height=20)

ggplot(data = abdata_bif_melted_all_bif, aes(x = as.factor(timepoint), y = log(value+0.00001))) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point() + geom_line(aes(group=Sample_ID),color='grey') + facet_grid(cols=vars(rx)) + stat_compare_means(paired=T,method='t.test')
ggsave('bifido_rx_overall_timepoint.pdf',width=5,height=4)

```

```{r}
#baseline multinomial associations

data = bind_rows(baseline_metaphlan,endpoint_metaphlan) %>% column_to_rownames('Sample_ID') 
#data = data %>% select(!all_of(grep('g__',colnames(data)))) 
data = data %>% select(all_of(grep('s__',colnames(data)))) %>% t

toremove = colnames(data) %>% strsplit('_') %>% map_chr(1) %>% table %>% data.frame %>% filter(Freq!=2) %>% select(colnames(.)[1]) %>% unlist %>% unname
data = data %>% data.frame%>% select(!(grep(paste(toremove, collapse="|"), colnames(data))))

# merge all data into one frame
metaphlan_s = data %>% t %>% data.frame %>% rownames_to_column('Sample_ID') %>% mutate(timepoint = strsplit(Sample_ID,'_') %>% map_chr(2)) %>% mutate(Sample_ID = strsplit(Sample_ID,'_') %>% map_chr(1))

abdata = left_join(metaphlan_s,metadata,by='Sample_ID') %>% filter(!is.na(RESP_STATUS))

abdata$RESP_STATUS = factor(abdata$RESP_STATUS,levels=c("NOCHANGE-TREATMENT","IMPROVED-PLACEBO","RESPONDER","NOCHANGE-PLACEBO","WORSE-PLACEBO","WORSE-TREATMENT"))

abdata_b = abdata %>% filter(timepoint=='baseline') 
abdata_e = abdata %>% filter(timepoint=='endpoint')

#microbes = abdata_b %>% select(grep('s__',colnames(abdata_b)))
#toremove = microbes[sapply(microbes, function(x) mean(x == 0) > 0.9)]  %>% colnames

#abdata_b= abdata_b %>% select(-all_of(toremove))
regcolumns = colnames(abdata_b)[grep('k__',colnames(abdata_b))]

#abdata_b = abdata_b %>% filter(RESP_STATUS!='NONRESP')
#abdata_e = abdata_e %>% filter(RESP_STATUS!='NONRESP')

#abdata_b = abdata_b %>% filter(RESP_STATUS == 'RESPONDER' | RESP_STATUS == 'NOCHANGE-TREATMENT')
#abdata_b$RESP_STATUS = as.numeric(factor(abdata_b$RESP_STATUS)) -1

output = list()
for(r in regcolumns){
  regout = multinom(data = abdata_b,family='binomial', RESP_STATUS ~ age+ (abdata_b[,r])) %>% tidy %>% mutate(microbial_var = r) %>% filter(term!='(Intercept)',term!='age')
  prevs = abdata_b %>% select(r,RESP_STATUS) %>% mutate(prev = if_else(get(r)>0,1,0)) %>% group_by(RESP_STATUS) %>% summarize(prev=sum(prev))
output[[r]] =left_join(regout,prevs,by=c('y.level'='RESP_STATUS'))
  }
output_merged = bind_rows(output) %>% mutate(adj=p.adjust(p.value,method = 'BY'))

for(val in unique(output_merged$y.level)){
  print(ggplot(data = output_merged %>% filter(y.level == val),aes(x=estimate,y=-log10(adj))) + geom_point() + xlim(-5,5) +ggtitle(unique(output_merged%>% filter(y.level == val)%>% select(y.level) %>% unique)) + ylab('-log10(qval)') + geom_label_repel(data = output_merged%>% filter(y.level == val) %>% arrange(adj) %>% filter(adj<0.001) %>% head(15),aes(label = microbial_var),box.padding   = 0.1, point.padding = 0.1,max.overlaps=100,size=3,segment.color = 'grey50')+geom_hline(yintercept = -log10(0.001)))
}

```

```{r}
# plot the most significant associations

setwd('~/Desktop/google_drive/My Drive/pds08/association_plots_species')

microbes_of_interest = output_merged %>% filter(adj<0.05) %>% arrange(desc(abs(prev))) %>% select(microbial_var) %>% unlist %>% unname %>% unique

for(m in microbes_of_interest){
ggplot(abdata_b %>% select(RESP_STATUS,all_of(grep(m,colnames(abdata_b)))),aes(x=RESP_STATUS,y=abdata_b[,m])) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_point()
  ggsave(paste(m,'_multinom.pdf',sep=''))
}

```

```{r}
#rulefit regression
rulefit_data = abdata %>% filter(timepoint=='baseline') %>% select(-c(d_bm1,d_bm2,d_bm3,bm_weekly,RESP_STATUS,RESP_V_NONRESP,IMPROVED,Randomization_Assignment,Seed_ID,id)) %>% select_if(~ !any(is.na(.))) %>% mutate_if(is.character,as.factor)
rulefit_data = Filter(function(x)(length(unique(x))>1), rulefit_data)

set.seed(42)
rulefit_output <- pre(d_bm_weekly ~ ., data = rulefit_data)

```

```{r}
#endpoint multinomial associations

output = list()
for(r in regcolumns){
  output[[r]] = multinom(data = abdata_e, RESP_STATUS ~ age+ log(abdata_e[,r]+0.00001),family='binomial') %>% tidy %>% mutate(microbial_var = r %>% strsplit('s__') %>% map_chr(2)) %>% filter(term!='(Intercept)',term!='age')
}
output_merged = bind_rows(output) %>% mutate(adj=p.adjust(p.value,method = 'BY'))

for(val in unique(output_merged$y.level)){
  print(ggplot(data = output_merged %>% filter(y.level == val),aes(x=estimate,y=-log10(adj))) + geom_point() + xlim(-5,5) +ggtitle(unique(output_merged%>% filter(y.level == val)%>% select(y.level) %>% unique)) + ylab('-log10(qval)') + geom_label_repel(data = output_merged%>% filter(y.level == val) %>% arrange(adj) %>% filter(adj<0.05) %>% head(15),aes(label = microbial_var),box.padding   = 0.1, point.padding = 0.1,max.overlaps=100,size=3,segment.color = 'grey50')+geom_hline(yintercept = -log10(0.05)))
  ggplot(abdata_b %>% select(RESP_STATUS,all_of(grep(m,colnames(abdata_b)))),aes(x=RESP_STATUS,y=abdata_b[,m])) + geom_boxplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}

```

```{r}
# parse and merge in card data
setwd('~/Desktop/google_drive/My Drive/pds08/')

card_files = list.files('card_genes')

output = list()
for(f in card_files){
  data = read.csv(paste('card_genes/',f,sep=''),sep='\t',header=T) %>% select(Drug.Class,Best_Hit_ARO,AMR.Gene.Family,Resistance.Mechanism)
  sample_id = strsplit(f,'_') %>% map_chr(1)
  timepoint = strsplit(f,'_') %>% map_chr(2)
  data$Sample_ID = sample_id
  data$timepoint = timepoint
  output[[f]] = data
}

merged_output = bind_rows(output) %>% mutate(timepoint = gsub('01','baseline',timepoint)) %>% mutate(timepoint = gsub('02','endpoint',timepoint))
merged_output_melted = melt(id = c('Sample_ID','timepoint'),merged_output) 
merged_output_melted_counted = merged_output_melted %>% group_by(Sample_ID,timepoint,variable,value) %>% count
merged_output_melted_counted_unique = merged_output_melted %>% unique %>% group_by(Sample_ID,timepoint,variable) %>% count

merged_output_melted_counted_unique_mdat = left_join(merged_output_melted_counted_unique,metadata)
merged_output_melted_counted_mdat = left_join(merged_output_melted_counted,metadata)
merged_output_melted_counted_mdat = left_join(merged_output_melted_counted_mdat,divs)

# variable level summaries
merged_output_melted_summary_overall = merged_output_melted %>% unique %>% group_by(Sample_ID,timepoint,variable) %>% count
merged_output_melted_summary_overall_mdat = left_join(merged_output_melted_summary_overall,metadata)

# unmelted data for regression
merged_output_mdat = left_join(merged_output,metadata)
merged_output_mdat = left_join(merged_output_mdat,divs)
```

```{r}
#  total card genes by type by resp non resp

# overall baseline/endpoint
ggplot(data = merged_output_melted_summary_overall_mdat %>% filter(timepoint == 'baseline',!is.na(RESP_STATUS)) ,aes(x=Sample_ID,y=n)) + geom_bar(stat='identity') + facet_grid(cols=vars(RESP_STATUS),rows=vars(variable),scales='free') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

ggplot(data = merged_output_melted_summary_overall_mdat %>% filter(timepoint == 'endpoint',!is.na(RESP_STATUS)) ,aes(x=Sample_ID,y=n)) + geom_bar(stat='identity') + facet_grid(cols=vars(RESP_STATUS),rows=vars(variable),scales='free') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

### RES MECHS 

# baseline counts
ggplot(data = merged_output_melted_counted_mdat %>% filter(timepoint=='baseline',variable == 'Resistance.Mechanism',!is.na(RESP_STATUS)),aes(x=Sample_ID,y=n)) + geom_bar(stat='identity') + facet_grid(cols=vars(RESP_STATUS),rows=vars(value),scales='free') + theme(axis.text.x = element_text(angle = 0, vjust = 0.5, hjust=1),axis.text.y = element_text(angle = 90, vjust = 0.5, hjust=1))

# endpoint counts
ggplot(data = merged_output_melted_counted_mdat %>% filter(timepoint=='endpoint',variable == 'Resistance.Mechanism',!is.na(RESP_STATUS)),aes(x=Sample_ID,y=n)) + geom_bar(stat='identity') + facet_grid(cols=vars(RESP_STATUS),rows=vars(value),scales='free') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+theme(axis.text.y = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r}
# association between different var types and resp status

amr_associations <- function(amr_data,metadata,diversity_data,colname, timeval){
   merged_output_melted = melt(merged_output,id.vars = c('Sample_ID','timepoint')) %>% filter(variable == colname) %>% select(-variable) %>% filter(timepoint == timeval) %>% select(-timepoint)
   merged_output_unmelted = merged_output_melted %>% dcast(Sample_ID ~ value)
   merged_output_unmelted_summed = merged_output_unmelted %>% column_to_rownames('Sample_ID') %>% as.data.frame %>% rowSums %>%as.data.frame() %>% rownames_to_column('Sample_ID')
   colnames(merged_output_unmelted_summed)[2]='n'
   regvals = merged_output_unmelted %>% select_if(is.numeric) %>% colSums %>% as.data.frame %>% filter(.>3) %>% rownames
   merged_output_unmelted = merged_output_unmelted %>% select(Sample_ID, all_of(regvals))
   data_sub = left_join(merged_output_unmelted,metadata)
   data_sub = left_join(data_sub,diversity_data %>% filter(timepoint == timeval) %>% select(-timepoint))
   output=list()
  for(c in regvals){
    # nonresp vs resp
    nr = glm(data = data_sub, RESP_V_NONRESP ~ data_sub[,c] + age,family='binomial') %>% tidy %>% mutate(variable = colname, type = c, depvar = 'RESP_V_NONRESP')
    # diversity/richness
    ri = glm(data = data_sub, richness ~ data_sub[,c] + age) %>% tidy %>% mutate(variable = colname, type = c,depvar = 'richness')
    sh = glm(data = data_sub, shannon ~ data_sub[,c] + age) %>% tidy %>% mutate(variable = colname, type = c,depvar = 'shannon')
    si = glm(data = data_sub, simpson ~ data_sub[,c] + age) %>% tidy %>% mutate(variable = colname, type = c,depvar = 'simpson')
    output[[c]] = bind_rows(nr,ri,sh,si)
  }
  overall = left_join(merged_output_unmelted_summed,metadata)
  overall = left_join(overall,diversity_data %>% filter(timepoint == timeval) %>% select(-timepoint))
    nr = glm(data = overall, RESP_V_NONRESP ~ n + age,family='binomial') %>% tidy %>% mutate(variable = colname, type = 'overall', depvar = 'RESP_V_NONRESP')
    # diversity/richness and gene counts
    ri = glm(data = overall, richness ~ n + age) %>% tidy %>% mutate(variable = colname, type = 'overall',depvar = 'richness')
    sh = glm(data = overall, shannon ~ n + age) %>% tidy %>% mutate(variable = colname, type = 'overall',depvar = 'shannon')
    si = glm(data = overall, simpson ~ n + age) %>% tidy %>% mutate(variable = colname, type = 'overall',depvar = 'simpson')
    output[['overall']] = bind_rows(nr,ri,sh,si)
  output = bind_rows(output) %>% filter(term != '(Intercept)',term!='age') %>% select(-term) %>% mutate(bh = p.adjust(p.value,method = 'BH'))
  return(output)
}

amr_associations(merged_output_mdat,metadata,divs,"Resistance.Mechanism",'baseline') %>% filter(bh<0.05)
amr_associations(merged_output_mdat,metadata,divs,"AMR.Gene.Family",'baseline') %>% filter(bh<0.05)
amr_associations(merged_output_mdat,metadata,divs,"Best_Hit_ARO",'baseline') %>% filter(bh<0.05)
amr_associations(merged_output_mdat,metadata,divs,"Drug.Class",'baseline') %>% filter(bh<0.05)
```


```{r}
ggplot(data = left_join(metadata,divs), aes(x=richness,fill=as.factor(RESP_STATUS))) + geom_histogram(bins=50)
```

